### PR TITLE
Metadata should not require compilation 

### LINKED ISSUE #5682: Metadata should not require compilation
Related: #3437

### PR DESCRIPTION
**Pull request by axic**
Closes #5682. Depends on #6139.

### CODE REVIEW

File: libsolidity/interface/CompilerStack.cpp [hunk 1]
```diff
@@ -97,6 +97,24 @@ void CompilerStack::setEVMVersion(EVMVersion _version)
 	m_evmVersion = _version;
 }
 
+void CompilerStack::setLibraries(std::map<std::string, h160> const& _libraries)
+{
+	solAssert(m_stackState < State::AnalysisSuccessful, "Set libraries after analysis.");
```
Comment (chriseth) [inline libsolidity/interface/CompilerStack.cpp]: We explicitly throw CompilerError for other cases of mis-use of the interface.

**Review by chriseth — CHANGES_REQUESTED** [review]
This at least requires a change in StandardCompiler.cpp:204 - it might be that @christianparpart already has the required change.

**Comment by axic** [comment]
Please comment on https://github.com/ethereum/solidity/issues/5682#issuecomment-460856179. We need to answer these questions. I'm not convinced it is the right direction.

> This at least requires a change in StandardCompiler.cpp:204

it actually isn't a problem in standardcompiler - in there parse/compile is called after all the settings.

File: libsolidity/interface/CompilerStack.cpp (hunk 1 above)
Comment (axic) [reply libsolidity/interface/CompilerStack.cpp]: This is a copy of `setEVMVersion` above.

**Comment by axic** [comment]
@christianparpart feel free to work on this / push your changes.

**Comment by christianparpart** [comment]
@chriseth @axic I've done the requested change. Too sad I did the same in my branch. But this one here looked more compled. Thx @axic for your aftershow-inspirations though. :-)

**Comment by chriseth** [comment]
Still needs a change in `isBinaryRequested` in StandardCompiler.cpp.

**Comment by axic** [comment]
> Still needs a change in `isBinaryRequested` in StandardCompiler.cpp.

Fixed, but that means we should definitely add a test.

**Comment by chriseth** [comment]
Yep, I agree. There should already be similar tests, if I remember correctly.

**Comment by axic** [comment]
Should be ready.

**Review by chriseth — DISMISSED** [review]

**Comment by axic** [comment]
The reason bytecodecomparison fails on travis is that it uses the 0.5.4 release and not the soljson from the PR:
```
solcVersion: 0.5.4+commit.9549d8ff.Emscripten.clang
packageVersion: 0.5.4
Version matching
```

**Comment by axic** [comment]
Rebased after #5959.

**Comment by axic** [comment]
@chriseth maybe if you want to check this, for some reason in those bytecodecomparison tests the contract is not compiled anymore (bytecode is not returned), despite bytecode is requested.

**Comment by chriseth** [comment]
It tries to compile a contract that has unimplemented features.

**Comment by chriseth** [comment]
Due to the changes in this PR, it can still generated valid metadata. The script though assumes that there is only an error if the 'contract' key is missing.

Will fix.

**Comment by axic** [comment]
> It tries to compile a contract that has unimplemented features.

I've seen that. But previously it didn't fail somehow when the metadata was requested?

**Comment by axic** [comment]
> Due to the changes in this PR, it can still generated valid metadata. The script though assumes that there is only an error if the 'contract' key is missing.

Ah, right. All these scripts...

**Review by chriseth — APPROVED** [review]

**Pull request merged**